[8.0] [ADD] project_task_category

martintorresvm opened the pull request:
New simple module that have m2o reference to project.category

Commit:
[ADD] project_task_category
New simple module that have m2o reference to project.category

elicoidal commented on project_task_category/README.rst:
```diff
@@ -0,0 +1,47 @@
+------------
+
+* Victor Martin <victor.martin@elico-corp.com>
+
```
Add my name

elicoidal commented on project_task_category/README.rst:
```diff
@@ -0,0 +1,47 @@
+Project Task Category
+=====================
+
+This module extends the functionality of Project to add an unique mandatory category on the task and support filter and group.
```
no more than 80char.

elicoidal commented on project_task_category/models/project_task_category.py:
```diff
@@ -0,0 +1,15 @@
+
+    task_categ_id = fields.Many2one(
+        'project.category',
+        'Task Category',
```
Category is enough

elicoidal replied:
It should be m2o with the model used in the tags

martintorresvm replied:
@elicoidal Maybe I made I mistake but I think actually it is.

elicoidal replied:
LGTM. It is indeed

elicoidal reviewed (changes requested):
Model link to be added

Commit:
[FIX] project_task_category
Fix Odoo internal field is already created by another module.
	modified:   project_task_category/models/project_task_category.py
	modified:   project_task_category/views/project_task_view.xml

Commit:
[FIX] project_task_category
Add others contributors and fix layout.

martintorresvm commented:
Please have a check @elicoidal

elicoidal commented on project_task_category/README.rst:
```diff
@@ -0,0 +1,49 @@
+
+ * have basic modules installed (project)
+
+Bug Tracker
```
Please add the Runbot

Commit:
[FIX] project_task_category
Add runbot and Usage section on Readme.
	modified:   project_task_category/README.rst

martintorresvm commented:
@elicoidal Could you approve the changes requested?

martintorresvm commented:
Message in Runbot is weird to me furthermore Travis is not failing.

2017-03-02 02:18:05,174 186 INFO openerp_test openerp.modules.module: module project_task_category: creating or updating database tables
2017-03-02 02:18:05,202 186 ERROR openerp_test openerp.sql_db: bad query: ALTER TABLE "project_task" ALTER COLUMN "categ_id" SET NOT NULL
Traceback (most recent call last):
File "/root/odoo-8.0/openerp/sql_db.py", line 234, in execute
res = self._obj.execute(query, params)
IntegrityError: column "categ_id" contains null values

2017-03-02 02:18:05,203 186 WARNING openerp_test openerp.models: WARNING: unable to set column categ_id of table project_task not null !
Try to re-run: openerp-server --update=module
If it doesn't work, update records and execute manually:
ALTER TABLE project_task ALTER COLUMN categ_id SET NOT NULL
Traceback (most recent call last):
File "/root/odoo-8.0/openerp/models.py", line 2670, in _auto_init
cr.execute('ALTER TABLE "%s" ALTER COLUMN "%s" SET NOT NULL' % (self._table, k))
File "/root/odoo-8.0/openerp/sql_db.py", line 158, in wrapper
return f(self, *args, **kwargs)
File "/root/odoo-8.0/openerp/sql_db.py", line 234, in execute
res = self._obj.execute(query, params)
IntegrityError: column "categ_id" contains null values

@elicoidal @dreispt @moylop260 @pedrobaeza Could you give a hand to understand why or if I could avoid this.

moylop260 commented:
@victormartinelicocorp 


The error [IntegrityError: column "categ_id" contains null values](https://travis-ci.org/OCA/project/jobs/206843658#L1033-L1038) means:
 - You are changing from not required to required a field, but there are old records with null values for this one

```python
class ProjectTask(models.Model):
    _inherit = 'project.task'
 
    categ_id = fields.Many2one(
        'project.category',
        'Category',
        required=True   # Add NOT NULL constraint for this field
    )
```

Note: Travis there are errors too but the log is not changing the state (because runbot use a regex to parser the log and travis use hard-coded blacklist... other MQT issue)
https://travis-ci.org/OCA/project/jobs/206843658#L1033-L1038

moylop260 commented:
The database template (with modules dependencies installed before to install `project/` modules) has data demo.

Then if you try create a new field with a NOT NULL constraint you will be the error because the old records don't have this field filled.

![screen shot 2017-03-01 at 11 33 21 pm](https://cloud.githubusercontent.com/assets/6644187/23494142/80a83d7c-fed6-11e6-8732-9506975c1baa.png)

Using the command `"ALTER TABLE project_task ALTER COLUMN categ_id SET NOT NULL"`
the output is: `ERROR:  column "categ_id" contains null values`

You can fix it:
Option A: Using a `default` value for `categ_id` field.
Option B: Remove `required` parameter for `categ_id` field.

My vote is for option B, because add new required fields you will have errors from other actions/methods that auto-create task without consider `categ_id`

martintorresvm commented:
> The database template (with modules dependencies installed before to install project/ modules) has data demo.
Then if you try create a new field with a NOT NULL constraint you will be the error because the old records don't have this field filled.

@moylop260 That is a really good point. 

@elicoidal Should we have one default category for Tasks? Which one? 

I would suggest do:
On PR #224 finish improve `project_categ` module and make the `Root Category for Tasks` mandatory

class ProjectProject(orm.Model):
     _columns = {
         'task_categ_id': fields.many2one(
             'project.category',
             required=True,
             default=lambda self: self.env.ref('project.project_category_01'),  
             'Root Category for Tasks'),

Then add default value `project.category` from Odoo Standard demo data.

And here the project.task `categ_id` should be mandatory with default value form project.project `task_categ_id`.

moylop260 commented:
This solution could add a major issue.
If the data is deleted then you will have a ugly traceback...

What about use required=False?

martintorresvm commented:
> This solution could add a major issue. If the data is deleted then you will have a ugly traceback... What about use required=False?

Understood but anyway we'd like to have an unique category linked to task independent from project and mandatory so I rather add default value just for Task and not modify project. Could be?

elicoidal commented:
I am  not sure to fully understand the issue. 
Categ_id field should be created by this module and is distinct from the tags. It should be specific for this module so we can do whatever we want with it and have it required.
This means that another non standard module has created this field, right?
@moylop260 @victormartinelicocorp

martintorresvm commented:
I think the problem is implicit with de demo data as it's used in runbot. 

Travis hasn't this issue.

Runbot first created DB with demo data with the standard Odoo modules installed.

Then over this DB try to install the modules within the PR.

So what it happens if you try to set a field as not null over a non empty table it's normal.

Many options are valid to solve this. One could be add default values 

@moylop260 @elicoidal

dreispt commented on project_task_category/models/project_task_category.py:
```diff
@@ -0,0 +1,15 @@
+    categ_id = fields.Many2one(
+        'project.category',
+        'Category',
+        required=True
```
Why do we need the Category to be always mandatory?
I can see the case for it to be mandatory depending on the Project...
How about having it mandatory only on the Task Form? That way it will be easier to customize more specific use cases (and to have it installed on databases already having Task data).

moylop260 commented:
Did you review my first comment?

If you see the log of travis you will see the same error
https://travis-ci.org/OCA/project/jobs/206843658#L1033-L1038

Commit:
Merge branch '8.0' of github.com:OCA/project into 8.0-add-project_task_category

Commit:
[FIX] project_task_category
Add required field at Form level
	modified:   project_task_category/models/project_task_category.py
	modified:   project_task_category/views/project_task_view.xml

martintorresvm commented:
> Model link to be added

@elicoidal Where I should add model link?


@moylop260 @dreispt I think I have done the necessary please could you have a look into my latest commit?

dreispt reviewed (dismissed):
OK to merge with squash.
It would be nice to add a note to the README stating that the Category field is mandatory.

Commit:
[FIX] project_task_category
Add on Readme the field Category is required.
	modified:   project_task_category/README.rst

Commit:
Merge branch '8.0' of github.com:OCA/project into 8.0-add-project_task_category

martintorresvm commented:
> OK to merge with squash.
It would be nice to add a note to the README stating that the Category field is mandatory.

@dreispt done

>   Model link to be added
@elicoidal Where I should add model link?

dreispt commented:
I just noticed that this is a new module.
But this feature is already implemented here: https://github.com/OCA/project/blob/8.0/project_categ/project_categ_model.py#L78

martintorresvm commented:
> I just noticed that this is a new module.
But this feature is already implemented here: https://github.com/OCA/project/blob/8.0/project_categ/project_categ_model.py#L78

@dreispt Apparently is the same feature.

But with `project_categ` only is possible to have one `task_categ_id` peer project and for all the task linked.

With this module is possible to have one `categ_id` peer task and not dependent from project.

cc @elicoidal

elicoidal commented:
@dreispt the idea of this module is to add a specific field Category which is not a tag (m2m).
The purpose of it is to be able to manage statistics or BI report based on this field, which we cannot do with the tags field.
With your module, as far as I can read, we have a related field which means all tasks of a project would have the same content for the category.
This is not the purpose of this module: allow different category for different tasks.

martintorresvm commented:
@dreispt Could you check my latest comment?

cc @elicoidal

dreispt commented:
@elicoidal @victormartinelicocorp Yes, you are correct, sorry for my confusion.

dreispt commented:
@elicoidal You are the blocking element now :wink:

elicoidal commented:
hehe not anymore :p
